### ISSUE TITLE
Reservation key is not passed to the scripts

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1074]
In GitLab by @ainagaya on Jul 6, 2023, 13:30

Hello @dbeltrankyl and @kinow,

### Autosubmit Version

4.0.0b

### Expid affected(If applicable)

Tried with a009 in the VM (vainilla workflow), and a06v.

### Summary

I was trying to use the `RESERVATION` key. I tried both in Lumi and MN4:

```
PLATFORMS:
   LUMI:
      RESERVATION: "secret-name-to-skip-queues"
      PARTITION: "standard"
      TYPE: slurm

```

And in both cases the `--reservation="secret-name-to-skip-queues"` does not appear in the `.cmd` files. I put it in the custom directives, but it might be useful to have this key, and also it's quite easy to make mistakes with the custom directives key since it might be overwritten in another place unintentionally (trying to use completely different custom directives). As you said it might be easy to solve it so I can try to do it myself with your guidance! 

Thanks,

PD: maybe we could also have a key to exclude nodes. For Lumi especially.

()

**Comment by kinow** [comment]
>And in both cases the `--reservation="secret-name-to-skip-queues"` does not appear in the `.cmd` files. I put it in the custom directives, but it might be useful to have this key, and also it's quite easy to make mistakes with the custom directives key since it might be overwritten in another place unintentionally (trying to use completely different custom directives). As you said it might be easy to solve it so I can try to do it myself with your guidance!

Good idea IMO.

If you want to have a go at this, I think a good place to start would be the `SlurmPlatform`: https://earth.bsc.es/gitlab/es/autosubmit/-/blob/master/autosubmit/platforms/slurmplatform.py

It extends the `ParamikoPlatform`, so worth having that one open in your IDE/editor too: https://earth.bsc.es/gitlab/es/autosubmit/-/blob/master/autosubmit/platforms/paramiko_platform.py

The `SlurmPlatform` has a `self.header` that's an instance of `SlurmHeader`: https://earth.bsc.es/gitlab/es/autosubmit/-/blob/master/autosubmit/platforms/headers/slurm_header.py

I had a quick look at the code and I **think** the platform object [will have](https://earth.bsc.es/gitlab/es/autosubmit/-/blob/ccd753b624f61f1627317ade0e8e3586a3645e0f/autosubmit/platforms/paramiko_submitter.py#L191) the `self.reservation` with the value that you specified in the configuration file. The issue could be that we are simply not using it :slight_smile: 

>PD: maybe we could also have a key to exclude nodes. For Lumi especially.

If that's something that can be set in a Slurm directive, then I think you should be able to craft some code that adds it, probably modifying the same files linked above.

Thanks @ainagaya !

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 6, 2023, 14:47

Hello, 

Thanks to both, 

to add that it probably needs to be included in the wrappers as well

**Comment by kinow** [comment]
In GitLab by @ainagaya on Jul 14, 2023, 09:30

Hi @kinow @dbeltrankyl ,

I'm trying to follow the steps explained in the MWT meeting (now I have PyCharm and my own AS installation!). I created a branch for this purpose (https://earth.bsc.es/gitlab/es/autosubmit/-/tree/fix-reservation). 

I added (copying from the existing ones) a `get_reservation_directive` in `platforms/headers/slurm_header` and in the `platforms/paramiko_platform` I added the conditional regarding the Reservation key: https://earth.bsc.es/gitlab/es/autosubmit/-/commit/6411568b3d03ab6a5a496ce3627041530f256556


When trying to do an INSPECT, seems like `job.parameters['RESERVATION']` does not exist. I'm a little lost with how this list is defined. Any hint? I'm still learning how the checkpoints and the debug mode work.


Thanks,

**Comment by kinow** [comment]
Hi @ainagaya !

Great progress! :100: 

TL;DR: I believe you have to set that value in the `Job` object.

Longer version:

I looked at the diff between your branch and `master`, then did a quick experiment, to see the traceback. I think this is what you had?

```bash
Traceback (most recent call last):
  File "/home/kinow/Development/python/workspace/autosubmit/autosubmit/autosubmit.py", line 1491, in inspect
    Autosubmit.generate_scripts_andor_wrappers(
  File "/home/kinow/Development/python/workspace/autosubmit/autosubmit/autosubmit.py", line 1574, in generate_scripts_andor_wrappers
    job_list.check_scripts(as_conf)
  File "/home/kinow/Development/python/workspace/autosubmit/autosubmit/job/job_list.py", line 2283, in check_scripts
    if not job.check_script(as_conf, self.parameters, show_logs):
  File "/home/kinow/Development/python/workspace/autosubmit/autosubmit/job/job.py", line 1658, in check_script
    template_content,additional_templates = self.update_content(as_conf)
  File "/home/kinow/Development/python/workspace/autosubmit/autosubmit/job/job.py", line 1520, in update_content
    template_content = self._get_template_content(as_conf, snippet, template)
  File "/home/kinow/Development/python/workspace/autosubmit/autosubmit/job/job.py", line 1535, in _get_template_content
    return self._get_paramiko_template(snippet, template)
  File "/home/kinow/Development/python/workspace/autosubmit/autosubmit/job/job.py", line 1544, in _get_paramiko_template
    current_platform.get_header(self), self.executable),
  File "/home/kinow/Development/python/workspace/autosubmit/autosubmit/platforms/paramiko_platform.py", line 1225, in get_header
    '%RESERVATION_DIRECTIVE%', self.header.get_reservation_directive(job))
  File "/home/kinow/Development/python/workspace/autosubmit/autosubmit/platforms/headers/slurm_header.py", line 164, in get_reservation_directive
    if job.parameters['RESERVATION'] == '':
KeyError: 'RESERVATION'
```

Tracebacks like this are very useful. If you look at the bottom, it shows `KeyError`. This indicates the key `RESERVATION` was not set in `job.parameters`.

The lines above contains the frames and last lines in the execution of each one. Looking at the most recent call, I see `slurm_header.py", line 164`.

I opened your branch on my PyCharm, then I pressed (note: I use Eclipse Keymap, because I used to code with Eclipse :bow: ) CTRL+SHIFT+R to open the menu to quickly navigate between files. Then I typed `slurm_header` and selected the option with the Python file.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/52495aff4b6794ac758d9a69a7212dc6/image.png)

Then I pressed `CTRL+L` to go to a line, and typed `164` and then ENTER.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/4812c01160145b637e56e13e821c6c4f/image.png)

Just above the function you added, I saw the similar one for `NUMTHREADS`. So I considered that since `NUMTHREADS` is working, there must be something done to `NUMTHREADS` that's not done to `RESERVATION`.

I copied `NUMTHREADS`, right-clicked on the `autosubmit` folder with the sources, and selected "Find in Files". Then I pasted NUMTHREADS, enabled the filter for Python files, and the match case option too.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/24341ba257964d109d9995f868c9e63e/image.png)

The third option (`parameters['NUMTHREADS'] = self.threads`) in the list appears to be where that value is set. I double-clicked on that entry, and then added the entry for `RESERVATION` at the end. However, I type just 
`parameters['RESERVATION'] = self.`, then I pressed `CTRL+SPACE` (this is to force auto-complete in the IDE, it was one of the most famous shortcuts in Eclipse... not sure what's the equivalent with the default PyCharm keymap -- go to settings and type keymap, you can view it there I think.)

The auto-complete did not bring `self.reservation`… I right-clicked on the `self._tasks` (a few lines above), "Find Usages".

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/9a2d58f4c3b62dc1b35d0377a6fd311d/image.png)

PyCharm classifies the usages found. At the bottom (for me) I see "Value write". This contains where this variable is being set. The one that shows under `__init__` appears to be the constructor setting the default value.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/a090f6d0b275cfcf5b2d6e1758371ea4/image.png)

You probably need to set some value for reservation there, something like `self._reservation = ''`?

The other entry contains where its value is set. I double-clicked on this one, then the IDE opened its declaration.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/53e96ba698f1bb3e7d3546e43a525eb6/image.png)

So again, now I right clicked on the function definition, "Find Usages".

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/5c9fc4cedf1bdf9938b62fda55ea8953/image.png)

The result listing contains two places where I think you will be able to find where `tasks` is set for the Job object, and where you will probably have to do the same for `self.reservation`.

Once you do that, in your `job.py` (after you also added the `reservation` in your constructor/`__init__`) now you should be able to finish the change adding `parameters['RESERVATION'] = self.reservation`.

And trying `autosubmit inspect` should (at least in theory) work at this point :fingers_crossed: 

Take your time to read this comment and see if that makes sense. And if you get stuck just ping us and we can help you.

-Bruno

p.s. the shortcuts, options, and the logic I used to troubleshoot the traceback may be different from what Dani would do. You will probably have some preferences on how to use PyCharm, so feel free to try different things if you'd like! :nerd:

**Comment by kinow** [comment]
In GitLab by @ainagaya on Jul 14, 2023, 10:55

Thanks for those amazing explanations @kinow !! I was able to follow your detailed instructions and get the right header: 

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/956168757a9ccae93e623ea017db19c3/image.png)

I already pushed the changes. I didn't get into the wrappers part that @dbeltrankyl was mentioning. It's not that urgent because when we get a reservation we don't queue so we don't need wrappers. Anyway at some point it should be implemented probably.


Next step is to create the merge request and do some testing, right?

Thanks again. Your explanations were really easy to follow :dancer:

**Comment by kinow** [comment]
That was really fast @ainagaya !!! :clap: 

>I already pushed the changes. Next step is to create the merge request and do some testing, right?

Correct. And add a unit test if possible :grimacing: (sometimes writing the test takes longer than writing the code changes, which is normal)

Thanks Aina!

**Comment by kinow** [comment]
In GitLab by @ainagaya on Jul 14, 2023, 11:16

Cool! I created the merge request. I'm looking into other merge requests to see what kind of test should I do, but I might need another one of your masterclasses :blossom:

**Comment by kinow** [comment]
We can definitely try to come up with some unit test next week (calling it a day soon due to extra hours :runner: )

The latest merge request with tests that worked on a similar part of the code was @manuel-g-castro 's MR for the header/tailer config, but that's AS 3.x - https://earth.bsc.es/gitlab/es/autosubmit/-/merge_requests/348

Maybe you can start looking at his AS3 test, and see if that could be used in your MR.

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Jul 14, 2023, 13:11

And the worst part is that the test that I have coded for AS 3.15 does not work on AS 4 :cry:. I mean, the test works, but the `self.job.create_script` on AS 4 does not create the script on the temporary folder, then the test always fails.

**Comment by kinow** [comment]
In GitLab by @ainagaya on Jul 24, 2023, 11:10

Took a look into Manuel's merge request but I'm still a bit lost here. Maybe tomorrow we can have a short conversation (I think you will be in the office @kinow )? Thanks !!

**Comment by kinow** [comment]
Sure thing, @ainagaya ! If it's not too complicated, you can also ask your question here or on Slack and we can try to solve it. Whichever way you prefer :slight_smile: :thumbsup:

**Comment by kinow** [comment]
In GitLab by @ainagaya on Jul 24, 2023, 14:49

It's all about testing protocol. Do they automatically run in the GitLab? Or you create an experiment locally and perform tests with it?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 27, 2023, 15:23

@ainagaya 

I think you can check 

from test/unit/test_job.py

Line: 291 test_job_script_checking_contains_the_right_default_variables

**Comment by kinow** [comment]
In GitLab by @ainagaya on Jul 28, 2023, 09:31

Just pushed it. I'm not really confident with it haha but it passed the test. Maybe one of you can double-check? Thanks!!!

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 17, 2023, 10:46

mentioned in commit f5677a50ffff8489b920f653ee68e4b9c1abaf4c
